Source/Core/Core/AchievementManager.cpp

```diff
@@ -914,11 +917,39 @@ void AchievementManager::Request(const rc_api_request_t* request,
       });
 }
 
+// Currently, when rc_client calls the memory peek method provided in its constructor (or in
+// rc_client_set_read_memory_function) it will do so on the thread that calls DoFrame, which is
+// currently the host thread, with one exception: an asynchronous callback in the load game process.
+// This is done to validate/invalidate each memory reference in the downloaded assets, mark assets
+// as unsupported, and notify the player upon startup that there are unsupported assets and how
+// many. As such, all that call needs to do is return the number of bytes that can be read with this
+// call. As only the CPU and host threads are allowed to read from memory, I provide a separate
+// method for this verification. In lieu of a more convenient set of steps, I provide MemoryVerifier
+// to rc_client at construction, and in the Load Game callback, after the verification has been
+// complete, I call rc_client_set_read_memory_function to switch to the usual MemoryPeeker for all
+// future synchronous calls.
+u32 AchievementManager::MemoryVerifier(u32 address, u8* buffer, u32 num_bytes, rc_client_t* client)
+{
+  auto& system = Core::System::GetInstance();
+  u32 ram_size = system.GetMemory().GetRamSizeReal();
+  if (address >= ram_size)
+    return 0;
+  return std::min(ram_size - address, num_bytes);
+}
+
 u32 AchievementManager::MemoryPeeker(u32 address, u8* buffer, u32 num_bytes, rc_client_t* client)
 {
   if (buffer == nullptr)
     return 0u;
   auto& system = Core::System::GetInstance();
+  // If this is called in some other thread, for example a callback, we want to make sure we return
+  // how many bytes we *could* read even if we can't, because rc_client will deactivate achievements
+  // and leaderboards as "unsupported" if the memory access is invalid.
+  if (!(Core::IsHostThread() || Core::IsCPUThread()))
+  {
+    ASSERT(false);
```

Comment:
It would be better to include a message, like "MemoryPeeker called from wrong thread".